crates/mvr-api/src/handlers/struct_definition.rs

```diff
@@ -0,0 +1,184 @@
+}
+
+pub struct StructDefinition;
+
+impl StructDefinition {
+    pub async fn resolve(
+        Path(type_name): Path<String>,
+        State(state): State<Arc<AppState>>,
+    ) -> Result<Json<Response>, ApiError> {
+        verify_input(&type_name)?;
+
+        let tags = bulk_resolve_definitions_impl(state, vec![type_name.clone()]).await?;
+
+        let tag = tags
+            .get(&type_name)
+            .ok_or(ApiError::BadRequest(format!("type not found: {type_name}")))?
+            .clone()
+            .ok_or(ApiError::BadRequest(format!("type not found: {type_name}")))?;
+
+        Ok(Json(Response {
+            type_tag: Some(tag),
+        }))
+    }
+
+    pub async fn bulk_resolve(
+        State(state): State<Arc<AppState>>,
+        Json(payload): Json<BulkRequest>,
+    ) -> Result<Json<BulkResponse>, ApiError> {
+        for type_name in payload.types.iter() {
+            verify_input(type_name)?;
+        }
+
+        let tags = bulk_resolve_definitions_impl(state, payload.types).await?;
+
+        Ok(Json(BulkResponse {
+            resolution: tags
+                .into_iter()
+                .map(|(k, type_tag)| (k, Response { type_tag }))
+                .collect(),
+        }))
+    }
+}
+
+async fn bulk_resolve_definitions_impl(
+    state: Arc<AppState>,
+    types: Vec<String>,
+) -> Result<HashMap<String, Option<String>>, ApiError> {
+    let names = types
+        .iter()
+        .map(|type_name| NamedType::parse_names(type_name))
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        .flatten()
+        .map(|name| {
+            let versioned_name = VersionedName::from_str(&name)?;
+            Ok(ResolutionKey(versioned_name))
+        })
+        .collect::<Result<Vec<_>, ApiError>>()?;
+
+    let parsed_name_addresses = state
+        .loader()
+        .load_many(names.clone())
+        .await?
+        .into_iter()
+        .map(|(k, v)| (k.0.to_string(), v))
+        .collect::<HashMap<_, _>>();
+
+    let mapping_ref = Arc::new(parsed_name_addresses);
+
+    let futures = types.into_iter().map(|type_name| {
+        let state = state.clone();
+        let mapping = mapping_ref.clone();
+
+        async move { resolve_definition(type_name, &mapping, &state).await }
+    });
+
+    Ok(try_join_all(futures).await?.into_iter().collect())
+}
+
+/// Given a `type_name`, we try to resolve the definition of that TypeTag.
+/// This is different from the type resolution API, which can resolve any type (primitives, generics, etc.),
+/// but requires the full-type to be valid (cannot do partial generic resolution).
+async fn resolve_definition(
+    type_name: String,
+    mapping: &HashMap<String, ObjectID>,
+    state: &AppState,
+) -> Result<(String, Option<String>), ApiError> {
+    let fixed_type_tag = NamedType::replace_names(&type_name, mapping).ok();
+
+    if fixed_type_tag.is_none() {
+        return Ok((type_name, None));
+    }
+
+    let correct_type_tag = fixed_type_tag.unwrap();
```

Comment:
this can be replaced by:

```
    let Ok(correct_type_tag) = NamedType::replace_names(&type_name, mapping) else {
        return Ok((type_name, None));
    };
```